File: RemoveDuplicateLinkedList.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Sep 24 11:21:55 2020

@author: kevin
"""

"""
This problem was recently asked by Amazon:
Given a sorted linked list of integers, remove all the duplicate elements in 
the linked list so that all elements in the linked list are unique.
Here's an example and some starter code:
"""
class Node:
  def __init__(self, value, next=None):
    self.value = value
    self.next = next

  def __repr__(self):
    return f"({self.value}, {self.next})"


def remove_dup(lst):
  # Fill this in.
  tmp = lst
  while(tmp.next != None):
      if(tmp.value == tmp.next.value):
          tmp.next = tmp.next.next
      else:
          tmp = tmp.next

lst = Node(1, Node(2, Node(2, Node(3, Node(3)))))

remove_dup(lst)
print(lst)
# (1, (2, (3, None)))

File: RemoveDuplicateLinkedList.java
/*
This problem was recently asked by Amazon:
Given a sorted linked list of integers, remove all the 
duplicate elements in the linked list so that all elements in the linked list are unique.
*/

class Node{
   int data;
   Node next = null;
   
   /**
   * Constructor to create a linked list node.
   *
   * @param data This is the value of the linked list node to be initialized
   */
   public Node(int data){
      this.data = data;
   }
   
   /**
   * Prints the node's value in an easy to read format.
   *
   * @return A string containing the node's value and the next node's value (null if next node is null)
   */
   public String toString(){
      return data + " -> " + next;
   }
   
   
   /**
   * Prints the node's value in an easy to read format.
   *
   * @param n The start of the linked list to process
   * @return The start of the linked list
   */
   public Node removeDuplicateNodes(Node n){
      Node prev = n;
      Node cur = n.next;
      
      while(cur != null){
         if(cur.data == prev.data){
            cur = cur.next;
            prev.next = cur;
         }else{
            prev = prev.next;
            cur = cur.next;
         }
      }
      
      return n;
   }
}

class RemoveDuplicateLinkedList{
   public static void main(String[] args){
      Node node = new Node(1);
      node.next = new Node(2);
      node.next.next = new Node(2);
      node.next.next.next = new Node(3);
      node.next.next.next.next = new Node(3);
      System.out.println(node);
      //1 -> 2 -> 2 -> 3 -> 3
      
      System.out.println(node.removeDuplicateNodes(node));
      ////1 -> 2 -> 3   
   }
}